feat: follow migrated legacy library content block

navinkarkera opened the pull request:
<!--

Note: Please refer to the Support Development Guidelines on the wiki page to consider backporting to active releases:
https://openedx.atlassian.net/wiki/spaces/COMM/pages/4248436737/Support+Guidelines+for+active+releases

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly readable.
If the linked information must be private (because it contains secrets), clearly label the link as private.

-->

## Description

Allows authors to migrate legacy library content block in course to library v2 if the source library was migrated. The block works with the new library and has the same functionality as the new ItemBankBlock.

This PR also fixes some issues with Problem Bank like `Add components` button not working in the preview page and iframe postMessage failure after syncing of component in the preview page.

https://github.com/user-attachments/assets/68017884-0cca-4fc8-9ab1-d30af1fd4fc0

**Latest update in UI**

https://github.com/user-attachments/assets/0edb1e13-71ed-4eb8-8de0-42baa7e21561


Requirements as per The LegacyLibraryContentBlock OLX must work forever.
* [x] The LegacyLibraryContentBlock OLX must work forever.
* L points at a legacy library, which has been migrated to a new V2 library.
    * [x] Block Behavior: Preview and settings editor should look just like a modern ItemBank block.
    * [x] Child Behavior: If updates are available (see below), they are shown per-child on the View page, using the migrated child blocks as the update source.
* L points at a legacy library, which has not been migrated.
    * [x] Block and Child Behavior: Use existing LegacyLibraryContentBlock interface behavior.
* L is None, malformed, or doesn't point to an existing or migrated legacy library:
    * [x] Block Behavior: Display a validation error. Syncing is not allowed. (existing behavior)
    * [x] Child Behavior: Children are displayed if they exist, but no option to sync is offered.
**Sync behavior**
* [x] The parent LegacyLibraryContentBlock's version does not equal the migrated source library version (indicating that the legacy library received updates before its migration happened)

Useful information to include:

- Which edX user roles will this change impact? "Course Author" and "Developer"

## Supporting information

* https://github.com/openedx/frontend-app-authoring/issues/2386
* `Private-ref`: https://tasks.opencraft.com/browse/FAL-4251
* Related PR: https://github.com/openedx/frontend-app-authoring/pull/2491

## Testing instructions

* Create a legacy library with some components.
* Add a legacy library content block in course unit.
* Now migrate the legacy library.
* Refresh the course unit page, a new alert should appear on top of the library content block.
* Click on migrate button in the alert and verify that the block is migrated to Problem bank and behaves like it.
* The children blocks have upstream field set and show sync button if the v2 library component is updated.

## Deadline

"None" if there's no rush, or provide a specific date or event (and reason) if there is one.

## Other information

Include anything else that will help reviewers and consumers understand the change.

- Does this change depend on other changes elsewhere?
- Any special concerns or limitations? For example: deprecations, migrations, security, or accessibility.
- If your [database migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) can't be rolled back easily.

navinkarkera committed:
feat: show item bank ui for migrated legacy library content

navinkarkera committed:
feat: migrate legacy content block to item bank block on view in studio

navinkarkera committed:
fix: duplicate and copy issues

openedx-webhooks commented:
Thanks for the pull request, @navinkarkera!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

navinkarkera committed:
refactor: migration location and add tests

navinkarkera committed:
fix: lint issues

navinkarkera committed:
fix: item bank and library content children view add button functionality

Newly added blocks from library in children view page of item bank block
and migrated library content block were not displayed automatically.

navinkarkera committed:
fix: lint issues

navinkarkera committed:
fix: lint issues

navinkarkera committed:
feat: only migrate if same version of library is migrated

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -296,12 +352,27 @@ def _set_validation_error_if_empty(self, validation, summary):
+        if self.is_source_lib_migrated_to_v2 and not self.is_migrated_to_v2:
+            # If the source library is migrated but this block still depends on legacy library
+            # Migrate the block by setting upstream field to all children blocks
+            self._v2_update_children_upstream_version()
```
@navinkarkera  I'm not sure this is the best approach. The point of this method is to render the block, not to modify it--sticking the migration here is surprising and auto-magical. It will also update the children when the block is rendered in LMS, which would either fail or be a no-op, and would be confusing either way. In general, we should not be modifying content except in response to the action of a Studio user.

I had [tried to outline an idea to follow library_content child references _without_ a child migration process](https://github.com/openedx/frontend-app-authoring/issues/2386), but since writing that, I've also come to dislike the idea, because it would break the simplifying assumption that we can just check `self.upstream` on any block to see its upstream link. So, **I like that your PR uses a migration child step**, but I think we need some explicit way to trigger it.

Here's an idea:
* When a legacy library_content block has (a) _unmigrated_ chidlren and (b) a _migrated_ source library, it gives the users the option to "sync" the latest content, _regardless of whether are new changes in the target library_.
* Syncing, in this case, will migrate the children.

It's a slight product behavior change. In particular, every legacy library content block would have a "sync available" as soon as an operator migrates its source library. But, I think it would be worth the simplified behavior information architecture. What do you think?

navinkarkera replied:
@kdmccormick This is not the final version, locally I have something like this:
<img width="1336" height="591" alt="image" src="https://github.com/user-attachments/assets/6ef961f8-4248-46ab-b831-63a1b4cd030c" />

Sorry for the confusion.

navinkarkera replied:
Once the author clicks on `Migrate` it looks like this:

<img width="1475" height="711" alt="image" src="https://github.com/user-attachments/assets/81af8ec9-9993-4c00-98c5-1829118108fb" />

I'll update the PR description once I push the changes.

kdmccormick replied:
Cool, thanks. I think this is a good approach. I'll ask precisely what the UI should be at the sync-up tomorrow.

kdmccormick replied:
looping UX into the discussion here: https://github.com/openedx/frontend-app-authoring/issues/2386#issuecomment-3372472237

navinkarkera committed:
refactor: migrate block on request

navinkarkera committed:
fix: component reload on migration

navinkarkera committed:
fix: tests

rpenido reviewed (approved):
LGTM :+1:
Thank you for your work, @navinkarkera !

- [x] I tested this using the instructions from the PR
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

bradenmacdonald commented on xmodule/library_content_block.py:
```diff
@@ -90,19 +91,19 @@ class LegacyLibraryContentBlock(ItemBankMixin, XModuleToXBlockMixin, XBlock):
         scope=Scope.settings,
     )
+    is_migrated_to_v2 = Boolean(
+        # This is a hidden field that stores whether children blocks are migrated to v2 i.e, have upstream set
+        display_name=_("Is Migrated to library v2"),
+        scope=Scope.settings,
+        default=False,
+    )
```
@navinkarkera @kdmccormick  I'm confused about this. I get that we need to continue supporting the legacy block for backwards compatibility. But I thought the migration step in this PR would actually change the block type (in the OLX) from `library_content` to `itembank`. What is the downside of doing that instead of adding this new hybrid `is_migrated_to_v2` legacy block combination ?

The code seems to imply this in some cases, e.g.

> Migrate this legacy block to new ItemBankBlock which uses library v2 blocks as children.

> The source library has been migrated to v2, this block can be migrated to Problem bank

kdmccormick replied:
Because we would lose student state, particularly the `selected` field, which keeps track of the children that have been picked for each user.

(and yes, I would reword those comments. It'd be better to say that we're switching the legacy library content block into a mode where behaves like an itembank)

navinkarkera replied:
@kdmccormick Thanks! I'll update the comments.

bradenmacdonald replied:
Makes sense, and yeah the comments confused me because they didn't match the behavior.

I guess at some point down the road, once we know a full migration has been forced, `library_content` will just become an alias for `itembank` ?

kdmccormick replied:
unless we force migration of _course_ content, which we typically don't do, then I think that we'll perpetually want to support this logic:

> if I am a library_content block and my source_library_id points at a migrated legacy library, then ask the user if I want to "update the reference" (i.e., point the children at the migration's target library blocks and fetch the `upstream_` fields)

That could be a very thin compatibility wrapper around ItemBankBlock. All the other aspects of the LegacyLibraryContentBlock (the editor, the child viewer, the block picker, the legacy syncing mechanism, etc.) could go away as soon as Legacy Libraries are removed.

bradenmacdonald replied:
Sounds good. Can we document that in the code directly?

navinkarkera replied:
Updated here: https://github.com/openedx/edx-platform/pull/37405/commits/f49a1d21a519991fa52207608dc2fd6eea5e7772

bradenmacdonald replied:
Thanks!

navinkarkera committed:
refactor: comments and message wordings

ChrisChV reviewed (approved):
Looks good! Thanks for the great work!

ChrisChV commented:
@kdmccormick it's ready for another review

navinkarkera committed:
refactor: update alert text

navinkarkera committed:
docs: add context

kdmccormick commented on cms/djangoapps/modulestore_migrator/api.py:
```diff
@@ -83,3 +88,26 @@ def get_migration_info(source_keys: list[CourseKey | LibraryLocator]) -> dict:
+            return str(
+                LibraryUsageLocatorV2(lib_key, block_type=block_type, usage_id=usage_id)  # type: ignore[abstract]
+            )
+        except (ValueError, TypeError):
```
Please catch exceptions that are more specific, and/or put fewer lines within the `try` clause. All sorts of things throw ValueError and TypeError; you could be unintentionally concealing bugs by catching like this.

kdmccormick commented on cms/djangoapps/modulestore_migrator/api.py:
```diff
@@ -83,3 +88,26 @@ def get_migration_info(source_keys: list[CourseKey | LibraryLocator]) -> dict:
+
+def get_target_block_usage_keys(source_key: CourseKey | LibraryLocator) -> dict[UsageKey | None, str | None]:
+    """
+    Get all target blocks for given list of source keys.
```
This does not say anything about what the keys are and what the value are--please be more descriptive.

kdmccormick commented on cms/djangoapps/modulestore_migrator/api.py:
```diff
@@ -83,3 +88,26 @@ def get_migration_info(source_keys: list[CourseKey | LibraryLocator]) -> dict:
     }
+
+
+def get_target_block_usage_keys(source_key: CourseKey | LibraryLocator) -> dict[UsageKey | None, str | None]:
```
This is saying that the keys of the return dictionary are of type `UsageKey | None`. Why would the key ever be `None`? That sounds like it'd be a bug.

kdmccormick commented on cms/djangoapps/modulestore_migrator/api.py:
```diff
@@ -83,3 +88,26 @@ def get_migration_info(source_keys: list[CourseKey | LibraryLocator]) -> dict:
+            # mypy thinks LibraryUsageLocatorV2 is abstract. It's not.
+            return str(
+                LibraryUsageLocatorV2(lib_key, block_type=block_type, usage_id=usage_id)  # type: ignore[abstract]
+            )
```
Why stringify before returning? OpaqueKeys carry useful structure and type information. In this case, and in almost all other cases, it's best to pas around structured objects, and only convert to a string when it's absolutely necessary for serialization or display.

kdmccormick commented on cms/djangoapps/modulestore_migrator/api.py:
```diff
@@ -83,3 +88,26 @@ def get_migration_info(source_keys: list[CourseKey | LibraryLocator]) -> dict:
+        'source__key', 'target__key', 'target__learning_package__key'
+    )
+
+    def construct_usage_key(row: tuple[UsageKey | None, str, str]) -> str | None:
```
the first item of this 3-tuple is not a `UsageKey | None`. it's a `str`. Regardless of that, you're not even using it. Why pass it in at all? Instead of passing the whole row in, you could desctructure the row (`for (source_key, target_local_key, target_learning_package_key) in query_set`) and just pass in the data you need.

kdmccormick commented on cms/djangoapps/modulestore_migrator/api.py:
```diff
@@ -83,3 +88,26 @@ def get_migration_info(source_keys: list[CourseKey | LibraryLocator]) -> dict:
+    def construct_usage_key(row: tuple[UsageKey | None, str, str]) -> str | None:
+        try:
+            lib_key = LibraryLocatorV2.from_string(row[2])
+            _, block_type, usage_id = row[1].split(':')
```
Please add a comment describing the structure of `row[1]` so that the reader can understand what is happing in this parsing step.

kdmccormick commented on xmodule/item_bank_block.py:
```diff
@@ -541,6 +513,27 @@ def format_block_keys_for_analytics(self, block_keys: list[tuple[str, str]]) ->
+        """
+        Validates the state of this ItemBankBlock Instance.
+        """
+        return self._validate()
```
I see you renamed `validate` to `_validate` so that LegacyLibraryContentBlock could add custom validation behavior. But this is a textbook use of inheritance:, is it not? Instead of renaming, I think you could keep the original method named `ItemBankMixin.validate`, and then have a `LegacyLibraryContentBlock.validate` method which overrides it and calls `super().validate(...)` in order to use the base behavior.

navinkarkera replied:
@kdmccormick Yes, but we want to only use `ItemBankMixin.validate()` when the legacy block has been migrated else we want to use its parents, i.e. `XBlock.validate()` method.

Updated to use `super()` for migrated blocks and `XBlock.validate()` for unmigrated. See ecc6a0b3d581

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -90,19 +91,24 @@ class LegacyLibraryContentBlock(ItemBankMixin, XModuleToXBlockMixin, XBlock):
     )
+    # This is a hidden field that stores whether child blocks are migrated to v2, i.e., whether they have an upstream.
+    # We cannot completely remove this block code until we force-migrate course content. Otherwise, we'll lose student
+    # data (such as selected fields), which tracks the children selected for each user.
```
We will never force-migrate course content

```suggestion
    # We can never completely remove the legacy library_content block; otherwise, we'd lose student
    # data, such as selected fields, which tracks the children selected for each user.
```

kdmccormick reviewed (commented):
Looking pretty solid as a whole, but I'd like to see some more attention to the nitty details of code quality, especially around correctness of type annotations.

I'm on vacation until Oct 15, so if another reviewer is available before then to approve this, no need to wait on my approval.

navinkarkera committed:
fix: component links not being created on migrating legacy blocks

navinkarkera committed:
fix: api docs and types

navinkarkera committed:
refactor: use inheritance and specific parent method call

navinkarkera committed:
fix: imports

navinkarkera committed:
fix: api typing

navinkarkera reviewed (commented):
@kdmccormick Thanks! Addressed all your comments.

navinkarkera committed:
fix: upstream_version check

kdmccormick commented on cms/djangoapps/contentstore/tasks.py:
```diff
@@ -1677,7 +1677,7 @@ def handle_update_xblock_upstream_link(usage_key):
         LOGGER.exception(f'Could not find item for given usage_key: {usage_key}')
         return
-    if not xblock.upstream or not xblock.upstream_version:
+    if not xblock.upstream or xblock.upstream_version is None:
```
unnecessary?

navinkarkera replied:
No, since we set `upstream_version = 0` while migrating legacy blocks under LegacyLibraryContent, `not xblock.upstream_version` evaluates to `True` and the Entity links are not created.

kdmccormick replied:
gotcha

kdmccormick commented on cms/djangoapps/modulestore_migrator/api.py:
```diff
@@ -83,3 +89,32 @@ def get_migration_info(source_keys: list[CourseKey | LibraryLocator]) -> dict:
+            return None
+        try:
+            # Example: xblock.v1:problem:e9eef38f5f4c49de943c83a2d5170211
+            _, block_type, usage_id = usage_key_str.split(':')
```
```suggestion
            _, block_type, block_id = usage_key_str.split(':')
```

`usage_id` is a synonym for `usage_key`. What this slug is is correctly called a `block_id` or, in OLX terms, an `url_name`.

kdmccormick commented on cms/djangoapps/modulestore_migrator/api.py:
```diff
@@ -83,3 +89,32 @@ def get_migration_info(source_keys: list[CourseKey | LibraryLocator]) -> dict:
+        'source__key', 'target__key', 'target__learning_package__key'
+    )
+
+    def construct_usage_key(lib_key_str: str, usage_key_str: str) -> LibraryUsageLocatorV2 | None:
```
what we're calling `usage_key_str` here is not actually a usage key, it's a publishable entity key. Could you call it `entity_key_str`, here and at the call site?

bradenmacdonald replied:
Also, there is already a helper function in the libraries API for this, so you don't need to implement `construct_usage_key` at all.

Use this instead:

https://github.com/openedx/edx-platform/blob/da2daf255eb6e03d0935e48bb565849d04a29e4f/openedx/core/djangoapps/content_libraries/api/libraries.py#L620-L631

You just need to change your `query_set` to load `target__component` instead of `target__key`.

navinkarkera replied:
@bradenmacdonald `target__component` returns the component id which is an integer as we are using `values_list` in the db query to avoid loading unnecessary data.

bradenmacdonald replied:
Isn't `__component` an object, not an ID? `__component_id` or `__component__pk` would be the ID.

If there's some way to access these key parts using the `component_type.name` and `local_key`, like the existing function is doing, I think that's better than encoding knowledge of how to split the string key at `:` here in this repo, which feels like more of a Learning Core Internal thing. Maybe `target__component__local_key` etc. in the values_list?

kdmccormick replied:
I agree with Braden's point about using the model fields rather than parsing the key by hand. Otherwise this PR LGTM.

navinkarkera replied:
Updated here: d31584fc3b89
Used `select_related` and removed `values_list`.

navinkarkera committed:
refactor: rename variables

navinkarkera committed:
refactor: parsing entity keys to usage_keys

kdmccormick reviewed (approved):
🚀

kdmccormick commented:
Hey @navinkarkera , next time that you squash and merge, can you replace the long list of intermediate commit messages with a single helpful description? The intermediate messages aren't helpful once the commits are squashed into one.

navinkarkera commented:
@kdmccormick Yes, will keep that in mind.
